### PR TITLE
chore: replace usages of AddressBook with Roster in SyncGossip

### LINKED ISSUE #16370: Replace AB with Roster: SyncGossip

### PR DESCRIPTION
**Pull request by anthony-swirldslabs**
**Description**:
As a part of the TSS initiative, and specifically the TSS-Roster project, we're refactoring the Platform code to replace usages of `AddressBook` with `Roster`.

This fix refactors the `SyncGossip`-related classes.

**Related issue(s)**:

Fixes #16370

**Notes for reviewer**:
Tests should pass.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by mxtartaglia-sl — APPROVED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterUtils.java [hunk 1]
```diff
@@ -60,6 +65,47 @@ public static X509Certificate fetchGossipCaCertificate(@NonNull final RosterEntr
...
 
+    /**
+     * Fetch a hostname (or a string with an IPv4 address) of a ServiceEndpoint
+     * at a given index in a given RosterEntry.
+     *
+     * @param entry a RosterEntry
+     * @param index an index of the ServiceEndpoint
+     * @return a string with a hostname or ip address
+     */
+    public static String fetchHostname(@NonNull final RosterEntry entry, final int index) {
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterUtils.java:76]: I think we should add tests for these util methods

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/crypto/CryptoArgsProvider.java [hunk 2]
```diff
@@ -51,22 +53,26 @@ public class CryptoArgsProvider {
...
     public static AddressBook createAddressBook(final int size) {
-        final AddressBook addresses = RandomAddressBookBuilder.create(Randotron.create())
+        final Roster roster = RandomRosterBuilder.create(Randotron.create())
                 .withSize(size)
                 .withWeightDistributionStrategy(WeightDistributionStrategy.BALANCED)
                 .build();
 
+        // We still use the keys injection mechanism from the EnhancedKeyStoreLoader and CryptoStatic,
+        // so we use an AddressBook here.
+        AddressBook addresses = RosterUtils.buildAddressBook(roster);
```
Comment (edward-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/crypto/CryptoArgsProvider.java:74]: I've created an issue to bookmark replacing the use of `AddressBook` in the crypto loading classes: https://github.com/hashgraph/hedera-services/issues/16402

**Review by edward-swirldslabs — DISMISSED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/SyncGossip.java [hunk 3]
```diff
@@ -216,9 +224,10 @@ public SyncGossip(
 
         reconnectThrottle = new ReconnectThrottle(reconnectConfig, platformContext.getTime());
 
-        networkMetrics = new NetworkMetrics(platformContext.getMetrics(), selfId, addressBook);
+        networkMetrics = new NetworkMetrics(platformContext.getMetrics(), selfId, roster);
         platformContext.getMetrics().addUpdater(networkMetrics::update);
 
+        final AddressBook addressBook = RosterUtils.buildAddressBook(roster);
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/SyncGossip.java:230]: It looks like it is used in the `ReconnectHelper`, in unmodified code.    @anthony-swirldslabs , do we have an issue created for the refactor inside the reconnect helper?   I think its a good idea to create the issues related to wherever we have to translate back to `AddressBook`.

[the remark this answers is not in the document]
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/SyncGossip.java:230]: @lpetrovic05 : this is to minimize the size of this particular PR. The Reconnect-related code will be refactored at a later time.

@edward-swirldslabs : Yes, we do have a reconnect issue: https://github.com/hashgraph/hedera-services/issues/14985

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/Utilities.java [hunk 4]
```diff
@@ -364,23 +362,24 @@ public static boolean hasAnyCauseSuppliedType(
...
-                .filter(address -> !address.getNodeId().equals(selfId))
-                .map(address -> new PeerInfo(
-                        address.getNodeId(),
-                        Objects.requireNonNull(address.getHostnameExternal()),
-                        Objects.requireNonNull(address.getSigCert())))
+        return roster.rosterEntries().stream()
+                .filter(entry -> entry.nodeId() != selfId.id())
+                .map(entry -> new PeerInfo(
+                        NodeId.of(entry.nodeId()),
+                        // Assume that the first ServiceEndpoint describes the external hostname,
```
[the remark this answers is not in the document]
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/Utilities.java:379]: I acknowledge that the implementation deviates from the HIP created in the past. However, the current implementation is a step forward to deprecating the presence of internal endpoints in public data structures which we all agreed is the right direction. I suggest we update the documentation/HIP/proto at a later time.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterUtils.java (hunk 1 above)
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterUtils.java:76]: I added tests.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterUtils.java [hunk 5]
```diff
@@ -146,4 +192,60 @@ public static RosterEntry getRosterEntry(@NonNull final Roster roster, final lon
...
         throw new RosterEntryNotFoundException("No RosterEntry with nodeId: " + nodeId + " in Roster: " + roster);
     }
+
+    /**
+     * Build an Address object out of a given RosterEntry object.
+     * @param entry a RosterEntry
+     * @return an Address
+     */
+    @NonNull
+    public static Address buildAddress(@NonNull final RosterEntry entry) {
```
[the remark this answers is not in the document]
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterUtils.java:204]: I marked them as `@Deprecated`.

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/crypto/CryptoArgsProvider.java (hunk 2 above)
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/crypto/CryptoArgsProvider.java:74]: Thanks.

**Review by lpetrovic05 — APPROVED** [review]

**Review by mxtartaglia-sl — APPROVED** [review]

**Review by edward-swirldslabs — APPROVED** [review]

**Pull request merged**